### PR TITLE
Support recordset annotations

### PR DESCRIPTION
**Pull request by hatched**
The jujubundlelib now returns position annotations which needed to be supported in the GUI.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

Set up a real environment (LXC is fine) with the latest code.
```
mkdir -p ~/charms/trusty
cd ~/charms/trusty 
bzr branch lp:~juju-gui/charms/trusty/juju-gui/trunk juju-gui
juju bootstrap local
juju deploy --repository=~/charms local:trusty/juju-gui
juju set juju-gui juju-gui-source="https://github.com/hatched/juju-gui.git support-recordset-annotations"
```

Drop this bundle on the canvas and the services should be positioned properly.

```YAML
series: precise
services: 
  haproxy: 
    charm: "cs:precise/haproxy-35"
    num_units: 3
    annotations: 
      "gui-x": "-161.5"
      "gui-y": "-402"
    to: 
      - "1"
      - "2"
      - "lxc:0"
  "swift-proxy": 
    charm: "cs:precise/swift-proxy-30"
    num_units: 1
    annotations: 
      "gui-x": "124.5"
      "gui-y": "-93"
    to: 
      - "2"
  "squid-reverseproxy": 
    charm: "cs:precise/squid-reverseproxy-9"
    num_units: 1
    annotations: 
      "gui-x": "-292.5"
      "gui-y": "-123"
    to: 
      - "lxc:swift-proxy/0"
relations: []
machines: 
  "0": 
    series: precise
    constraints: "arch=amd64 cpu-cores=1 cpu-power=100 mem=1740 root-disk=8192"
  "1": 
    series: precise
    constraints: "arch=amd64 cpu-cores=1 cpu-power=100 mem=1740 root-disk=8192"
  "2": 
    series: precise
    constraints: "arch=amd64 cpu-cores=1 cpu-power=100 mem=1740 root-disk=8192"
```

File: app/utils/bundle-importer.js [hunk 1]
```diff
@@ -489,6 +493,21 @@ YUI.add('bundle-importer', function(Y) {
...
+
+    /**
+      Executes the setAnnotations method call
+
+      @method _execute_setAnnotations
+      @param {Object} record The setAnnotations record.
+      @param {Function} next The method to trigger the executor to move
+        on to the next record.
+    */
+    _execute_setAnnotations: function(record, next) {
```
Comment (mitechie) [inline app/utils/bundle-importer.js]: do we need/have any test that annotations are passed/working as expected?

**Comment by mitechie** [comment]
I'd like to check on a way to test annotations work in the new world to make sure we're good backwards compatibility wise before +1'ing this.

File: app/utils/bundle-importer.js (hunk 1 above)
Comment (hatched) [inline app/utils/bundle-importer.js]: None of these executors have any tests. They were in too much flux and buggy to write tests on their implementations. I have one more bug to fix then there will be a tests branch.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2500/
Test FAILed.

File: app/utils/bundle-importer.js [hunk 2]
```diff
@@ -489,6 +493,21 @@ YUI.add('bundle-importer', function(Y) {
...
+      var entityName = record[record.args[0].replace(/^\$/, '')].get('id');
+      var service = this.db.services.getById(entityName);
```
Comment (makyo) [inline app/utils/bundle-importer.js]: Machines can have annotations, too, in the v4 format.  We may want to check that the requires/args[0] is an addService type.

**Comment by mitechie** [comment]
ok, a hesitant :+1: please make sure we've got a card on the follow up tests and that annotations are noted to be something tested there as we've had a lot of layout breakages over time and this makes me nervous.

File: app/utils/bundle-importer.js (hunk 2 above)
Comment (hatched) [inline app/utils/bundle-importer.js]: Good idea - we don't support annotations in the GUI for machines.

**Comment by makyo** [comment]
:+1:+ QA okay

**Comment by hatched** [comment]
Thanks for the reviews and qa! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2502/
Test PASSed.

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/872

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/875

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**